lib/semantics/check-do.cc

```diff
@@ -64,6 +64,76 @@ class DoConcurrentBodyEnforce {
   template<typename T> bool Pre(const T &) { return true; }
   template<typename T> void Post(const T &) {}
 
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+
+  // C1140 -- Can't deallocate a polymorphic entity in a DO CONCURRENT.
+  // Deallocation can be caused by exiting a block that declares an allocatable
+  // entity, assignment to an allocatable variable, or an actual DEALLOCATE
+  // statement
+  //
+  // Note also that the deallocation of a derived type entity might cause the
+  // invocation of an IMPURE final subroutine.
+
+  // Deallocation caused by block exit
+  void Post(const parser::BlockConstruct &blockConstruct) {
+    const auto &endBlockStmt{
+        std::get<parser::Statement<parser::EndBlockStmt>>(blockConstruct.t)};
+    const Scope &blockScope{context_.FindScope(endBlockStmt.source)};
+    const Scope &doScope{context_.FindScope(doConcurrentSourcePosition_)};
+    if (DoesScopeContain(&doScope, blockScope)) {
+      for (auto &pair : blockScope) {
+        Symbol &symbol{*pair.second};
+        if (IsOrContainsPolymorphicComponent(symbol) && IsAllocatable(symbol) &&
+            !symbol.attrs().test(Attr::SAVE)) {
+          context_.SayWithDecl(symbol, endBlockStmt.source,
+              "Deallocation of a polymorphic entity caused by block"
+              " exit not allowed in DO CONCURRENT"_err_en_US);
+        }
+        // TODO: Check for deallocation of a variable with an IMPURE FINAL
+        // subroutine
+      }
+    }
+  }
+
+  // Deallocation caused by assignement
+  void Post(const parser::AssignmentStmt &stmt) {
+    const auto &variable{std::get<parser::Variable>(stmt.t)};
+    if (const Symbol * symbol{GetLastName(variable).symbol}) {
+      if (IsOrContainsPolymorphicComponent(*symbol)) {
```

Comment:
Should have the allocatable check also.

Reply:
D'oh!  Good catch.  I'll add a test for the non-error case, too.